pkg/console/service/service.go

```diff
@@ -0,0 +1,282 @@
+		return
+	}
+
+	vm, err := s.kubevirtClient.VirtualMachine(params.namespace).Get(params.name, &metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			_ = response.WriteError(http.StatusNotFound, fmt.Errorf("VirtualMachine does not exist: %w", err))
+			return
+		}
+		_ = response.WriteError(http.StatusInternalServerError, fmt.Errorf("error getting VirtualMachine: %w", err))
+		return
+	}
+
+	resourceName := vm.Name + "-vnc-access"
+	err = s.createResources(request.Request.Context(), resourceName, vm)
+	if err != nil {
+		_ = response.WriteError(http.StatusInternalServerError, fmt.Errorf("error creating resources: %w", err))
+		return
+	}
+
+	token, err := s.requestToken(request.Request.Context(), vm.Namespace, resourceName, params.duration)
+	if err != nil {
+		_ = response.WriteError(http.StatusInternalServerError, fmt.Errorf("failed to request token: %w", err))
+		return
+	}
+
+	_ = response.WriteAsJson(&v1alpha1.TokenResponse{
+		Token: token,
+	})
+}
+
+func (s *service) VncHandler(_ *restful.Request, response *restful.Response) {
+	_ = response.WriteError(http.StatusGone, fmt.Errorf(
+		"/vnc endpoint was removed",
+	))
+}
+
+type tokenRequestParams struct {
+	namespace string
+	name      string
+	duration  time.Duration
+}
+
+func readTokenRequestParameters(request *restful.Request) (*tokenRequestParams, error) {
+	namespace := request.PathParameter("namespace")
+	name := request.PathParameter("name")
+	if namespace == "" || name == "" {
+		return nil, fmt.Errorf("namespace and name parameters are required")
+	}
+
+	duration := 10 * time.Minute
+	durationParam := request.QueryParameter("duration")
+	if durationParam != "" {
+		var err error
+		duration, err = time.ParseDuration(durationParam)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse duration: %w", err)
+		}
+	}
+
+	return &tokenRequestParams{
+		namespace: namespace,
+		name:      name,
+		duration:  duration,
+	}, nil
+}
+
+func (s *service) createResources(ctx context.Context, name string, vmMeta metav1.Object) error {
+	const appLabelValue = "vm-console-proxy"
+
+	namespace := vmMeta.GetNamespace()
+	commonLabels := map[string]string{
+		AppKubernetesNameLabel:      appLabelValue,
+		AppKubernetesPartOfLabel:    appLabelValue,
+		AppKubernetesManagedByLabel: appLabelValue,
+		AppKubernetesComponentLabel: appLabelValue,
+	}
+
+	vmOwnerRef := metav1.OwnerReference{
+		APIVersion: kubevirt.GroupName,
+		Kind:       "VirtualMachine",
+		Name:       vmMeta.GetName(),
+		UID:        vmMeta.GetUID(),
+	}
+
+	serviceAccount, err := createOrUpdate[*core.ServiceAccount](
+		ctx,
+		name,
+		namespace,
+		s.kubevirtClient.CoreV1().ServiceAccounts(namespace),
+		func(foundObj *core.ServiceAccount) {
+			foundObj.Labels = commonLabels
+			foundObj.OwnerReferences = []metav1.OwnerReference{vmOwnerRef}
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create service account: %w", err)
+	}
+
+	role, err := createOrUpdate[*rbac.Role](
```

Comment:
Maybe instead of creating this role for each token we could make it a ClusterRole that is deployed with vm-console-proxy and drop the privilege to create new roles from vm-console-proxy?

Reply:
That is exactly what we don't want. A token should only allow access to a single VM.

Reply:
ACK